#15 16/chore/implement dto and dummy endpoint

YoshikiKubotani opened the pull request:
## Issue URL

#15 #16 

## Change overview

- Add new endpoints for summarization with dummy controllers and use cases
  - As described [here](https://fastapi.tiangolo.com/ja/tutorial/request-files/#what-is-form-data), it cannot be done to send files with json requests at the same time due to the HTTP protocol limitation
  - I implemented two workarounds for that based on [Stackoverflow (Method2)](https://stackoverflow.com/questions/65504438/how-to-add-both-file-and-json-body-in-a-fastapi-post-request) - (a) and on [GitHub Issues](https://github.com/tiangolo/fastapi/issues/2387#issuecomment-906761427) - (b)
  - In (a), non-file data are given by query parameters. But specifying query parameters in the POST process does not seem to be a standard solution, so it might be best to avoid it
    - [httpx](https://www.python-httpx.org/quickstart/) does not mention specifying the params argument in POST (actually, it generates errors)
    - The [FastAPI documentation](https://fastapi.tiangolo.com/ja/tutorial/testing/#_5) also says to write query parameters directly
    - Also, this method has a problem with the OpenAI API key appearing in the URL
  - In (b), non-file data is given from the client as form data filled out in JSON type
- Add necessary Data Transfer Object(DTO)s as pydantic models
- Update mypy version from 1.2.0 to 1.3.0
- Implement unit tests for endpoints
  - Unit tests are also done both on (a) and (b) mentioned above

## How to test

### Endpoint Unit Test

Run the following command.

```bash
# Run the following under Crux/environment/cpu
docker compose exec backend make test
```

### Architecture Check

Check the features separation is reasonable and follows [S.O.L.I.D Principles](https://medium.com/backticks-tildes/the-s-o-l-i-d-principles-in-pictures-b34ce2f1e898)

### Type Check

Run the following command.

```bash
# Run the following under Crux/environment/cpu
docker compose exec backend make mypy
```

## Note for reviewers

- If you explicitly wrap the endpoint returns in the `Response` class, the dynamic type checking by the pydantic model specified in response_model will not work. So I fixed the `Response` class in `default_response_class` and returned data as dictionary type.
- I tried to make the basis of the backend software architecture, but the folder structure remained unchanged under `backend/src`.  Please work on that task in #19 

## Questions

1. To solve the mypy error, I used pydantic models in parts other than `routers.py`, but is that correct?
2. The test is only passed when a fixed string (e.g., "This is a normal summary.") is returned from `controller.py`. When the logic is static, like Owan, the test can be configured to pass even after the logic is implemented. However, how should we configure the test for our case, where the solution changes dynamically (as the summary has a large variance)?
3. HTTP is supposed to be a stateless protocol, unlike FTP, etc. Cannot we send `multipart/form-data` and `application/json` consecutively as separate POST processes instead of simultaneously?

YoshikiKubotani committed:
:sparkles: Add an endpoint for pdf summarization

YoshikiKubotani committed:
:sparkles: Specify response class explicitly by `default_response_class` argment to avoid the extra overhead caused by passing the return content through the `jsonable_encoder`

See also https://fastapi.tiangolo.com/advanced/custom-response/#use-orjsonresponse

YoshikiKubotani committed:
:package: Add the `orjson` package

YoshikiKubotani committed:
:sparkles: Add a few more summary formats

YoshikiKubotani committed:
:sparkles: Change handling of PDF data given as user input

YoshikiKubotani committed:
:art: Separate DTO objects from routing

YoshikiKubotani committed:
:sparkles: Update the output format of DTO

YoshikiKubotani committed:
:bug: Fix bugs

YoshikiKubotani committed:
:package: Add the `python-multipart` package to use Form data in FastAPI

YoshikiKubotani committed:
:sparkles: Add domain/interface templates for dummy returns

YoshikiKubotani committed:
:building_construction: Add an empty core logic file for summarization

YoshikiKubotani committed:
:art: Rename a controller method

YoshikiKubotani committed:
:bug: Make sure to work pydantic type check

YoshikiKubotani committed:
:sparkles: Update the way of posting files with parameters

YoshikiKubotani committed:
:package: Update mypy version from 1.2.0 to 1.3.0

YoshikiKubotani committed:
:bug: Add quates to remove a mypy error massage

We get `pyproject.toml: [mypy]: python_version: Python 3.1 is not supported (must be 3.4 or higher). You may need to put quotes around your Python version` when doing mypy type check without quates

YoshikiKubotani committed:
:sparkles: Update types

YoshikiKubotani committed:
:bug: Make sure to return only summary sentences from `SummaryController`

YoshikiKubotani committed:
:package: Add the httpx package for using testclient

YoshikiKubotani committed:
:test_tube: Add the health check test

YoshikiKubotani committed:
:test_tube: Add  the summary test

YoshikiKubotani committed:
:art: Run lint and make the code format better

gatheluck commented:
@YoshikiKubotani It looks like CI still raise error due to code format? Can you fix this?

YoshikiKubotani commented:
@gatheluck 
That's because I added the basic template of the main logic under `backend/src/core_logic.py`, while it's not called.
and flake8 raises errors when functions and variables are defined but not used.

I think it's okay to ignore that since it would help us when we try to fully implement the logic in the future.

I've checked the type check and code format test were successfully passed.

gatheluck commented on applications/backend/src/dto.py:
```diff
@@ -0,0 +1,45 @@
+    title: str
+    author: str
+    key_figure: str
+    summary_text: str
```
```suggestion
    summary_text: FormatNormal | FormatThreePoint | FormatOchiai | FormatCVPaper
```
Well... To me, type of `summary_text` should be like above...?

YoshikiKubotani committed:
:package: Add necessary packages for general use of langchain

YoshikiKubotani committed:
:wrench: Update `.gitignore`

YoshikiKubotani committed:
:sparkles: Implement the simple top-k relevant chunks retriever

YoshikiKubotani committed:
:sparkles: Implement the code for summarizing the chunk-wise summaries

Implemented based on the tutorial [here](https://python.langchain.com/en/latest/modules/chains/index_examples/summarize.html)

YoshikiKubotani committed:
:art: Rename a file

YoshikiKubotani committed:
:sparkles: Implement the code for generating the three-point summary from the chunk-wise summaries

YoshikiKubotani committed:
:art: Rename the model name

YoshikiKubotani committed:
:speech_balloon: Remove unused comments

YoshikiKubotani committed:
:art: Make sure that the controller sends back the data in the `UserOut` form

YoshikiKubotani committed:
:bug: Fix type annotations

YoshikiKubotani committed:
:pencil2: Fix typo

YoshikiKubotani committed:
:test_tube: Update test code

YoshikiKubotani committed:
:art: Add FAISS import for overall uniformity

YoshikiKubotani committed:
:sparkles: Implement the code for generating study strength in ochiai format

YoshikiKubotani committed:
:sparkles: Add field description to dto objects

YoshikiKubotani committed:
:art: Build up the temp folder structure

YoshikiKubotani committed:
:sparkles: Implement ochiai format summary with vector search

YoshikiKubotani committed:
:sparkles: Implement multi-step approach to generate contribution summary of the ochiai format

YoshikiKubotani committed:
:sparkles: Add the mathpix PDF loader

YoshikiKubotani committed:
:sparkles: Update a template

YoshikiKubotani committed:
:sparkles: Implement latex splitter

YoshikiKubotani committed:
:package: Use the latest langchain library

YoshikiKubotani committed:
:sparkles: Add the updated version of contribution generation in the ochiai format into `all_ochi_related_qa.py`

YoshikiKubotani committed:
:art: Rename the file of generating ochiai format summary

YoshikiKubotani committed:
:sparkles: Make sure to add paper acronym to the output file names

YoshikiKubotani committed:
:sparkles: Update latex splitter

YoshikiKubotani committed:
:sparkles: Implement custom mathpix loader

YoshikiKubotani committed:
:art: Rename splitter variable name

YoshikiKubotani committed:
:sparkles: Use `LatexSplitter` in ochiai format generation

YoshikiKubotani committed:
:sparkles: Update the code for normal summary generation

YoshikiKubotani committed:
:sparkles: Update ochiai format code to include abstract summary

YoshikiKubotani committed:
:art: Rename the file for ochiai format summary generation

YoshikiKubotani committed:
:art: Rename english template

YoshikiKubotani committed:
:sparkles: Add summary generation using Japanese prompts

YoshikiKubotani committed:
:sparkles: Add summary generation in japanese using English prompt

YoshikiKubotani committed:
:art: Run `make flake8`

YoshikiKubotani committed:
:sparkles: Use nakatsuka prompt

YoshikiKubotani committed:
:sparkles: Update ochiai format sumamry in japanese

YoshikiKubotani committed:
:sparkles: Add original latex parser

YoshikiKubotani committed:
:art: Remove unnecessary parts

YoshikiKubotani committed:
:art: Format code

YoshikiKubotani committed:
:wastebasket: Remove personal developing folder

YoshikiKubotani committed:
:building_construction: Transport completed code from `CVPR2023_Survey`

YoshikiKubotani committed:
Merge branch '#15-16/chore/implement-dto-and-dummy-endpoint' into dev-botany

YoshikiKubotani committed:
:art: Apply the basic clean architecture to the old files

YoshikiKubotani committed:
:art: Rename the controller for summarization

YoshikiKubotani committed:
:package: Update `pydantic` and `fastapi` version to the latest

`langchain=0.0.229` does not support `pydantic=2.0.0` yet, but the package list was changed for the future support

YoshikiKubotani committed:
:sparkles: Update summarize endpoint

YoshikiKubotani committed:
:building_construction: Add the gateway for RDB

YoshikiKubotani committed:
:sparkles: Update summary controller

YoshikiKubotani committed:
:art: Move mathpix loader under the usecase directory

YoshikiKubotani committed:
:sparkles: Add the io handler for papers

YoshikiKubotani committed:
:art: Rename `summarize_paper.py`

YoshikiKubotani committed:
:art: Move `loader.py`

YoshikiKubotani committed:
:sparkles: Update to process one file

YoshikiKubotani committed:
:sparkles: Add the usecase for pdf parser with mathpix

YoshikiKubotani committed:
:bug: Fix router arguments

YoshikiKubotani committed:
:bug: Fix bug related to Mathpix API

YoshikiKubotani committed:
:sparkles: Add necessary dtos

YoshikiKubotani committed:
:sparkles: Add prompts

YoshikiKubotani committed:
:sparkles: Add the core logic for summarization

YoshikiKubotani committed:
:sparkles: Update to work properly

YoshikiKubotani committed:
:wastebasket: Remove unnecessary files

YoshikiKubotani committed:
:wrench: Increase the timeout seconds

YoshikiKubotani committed:
:bug: Fix a minor bug

YoshikiKubotani committed:
:sparkles: Add dummy db

YoshikiKubotani committed:
↩️ Revert to use pydantic=1.10.10

YoshikiKubotani committed:
:art: Run `make format`

YoshikiKubotani committed:
:art: Refactor code to pass `make flake8`

YoshikiKubotani committed:
:package: Add `type-requests` to install stub files for the request package

YoshikiKubotani committed:
:art: Rename a variable

YoshikiKubotani committed:
:art: Fix `mathpix_loader.py` to pass mypy check

YoshikiKubotani committed:
:art: Remove `# flake8: noqa` from `ochiai_format_summarizer.py`

YoshikiKubotani committed:
:art: Fix `ochiai_format_summarizer.py` to pass mypy check

Changed `self.vectorstore["all"].docstore._dict(abstract_docstore_id)` to `self.vectorstore["all"].docstore.search(abstract_docstore_id)` due to the following reason.

`self.vectorstore` is instantiated in [`applications/backend/src/usecase/summary_handler.py`](https://github.com/cvpaperchallenge/Crux/blob/9c4f31f2607a191eaf6c2ee6eed829abbfca4169/applications/backend/src/usecase/summary_handler.py#L65).
This calls

1. [`from_documents`](https://github.com/hwchase17/langchain/blob/c4ece52dac8fb93c38fe818d5f1b006d29539409/langchain/vectorstores/base.py#L392) method of `vectorstores/base.py`
2. [`from_texts`](https://github.com/hwchase17/langchain/blob/c4ece52dac8fb93c38fe818d5f1b006d29539409/langchain/vectorstores/faiss.py#L542) method
3. [`__from`](https://github.com/hwchase17/langchain/blob/c4ece52dac8fb93c38fe818d5f1b006d29539409/langchain/vectorstores/faiss.py#L494) method of `vectorstores/faiss.py`

to instantiate the `FAISS` class [using `InMemoryDocstore`](https://github.com/hwchase17/langchain/blob/c4ece52dac8fb93c38fe818d5f1b006d29539409/langchain/vectorstores/faiss.py#L531).

`InMemoryDocstore` has [`search` method](https://github.com/hwchase17/langchain/blob/c4ece52dac8fb93c38fe818d5f1b006d29539409/langchain/docstore/in_memory.py#L29) to search the target `Document` from the vectorstore using docstore ids.

YoshikiKubotani committed:
:art: Remove unused pydantic model

YoshikiKubotani committed:
:art: Move `SummaryConfigDTO`

YoshikiKubotani committed:
:art: Fix `summary_handler.py` to pass mypy check

YoshikiKubotani committed:
:art: Fix `mathpix_pdf_parser.py` to pass mypy check

YoshikiKubotani committed:
:art: Fix `summary_controller.py` to pass mypy check

YoshikiKubotani committed:
:art: FIx `router.py` to pass mypy check

YoshikiKubotani committed:
:memo: Update docstrings

YoshikiKubotani committed:
:speech_balloon: Comment out the old test code

YoshikiKubotani committed:
Merge pull request #24 from cvpaperchallenge/dev-botany

Port the content implemented in the `CVPR_2023_Survey` repository

gatheluck reviewed (approved):
Let's merge this too if CI runs successfully.
